server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SmallChangesetQualityGateSpecialCase.java

```diff
@@ -0,0 +1,82 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.computation.task.projectanalysis.step;
+
+import java.util.Collection;
+import javax.annotation.Nullable;
+import org.sonar.api.measures.CoreMetrics;
+import org.sonar.server.computation.task.projectanalysis.component.Component;
+import org.sonar.server.computation.task.projectanalysis.measure.Measure;
+import org.sonar.server.computation.task.projectanalysis.measure.MeasureRepository;
+import org.sonar.server.computation.task.projectanalysis.metric.MetricRepository;
+import org.sonar.server.computation.task.projectanalysis.qualitygate.EvaluationResult;
+
+import static java.util.Arrays.asList;
+
+public class SmallChangesetQualityGateSpecialCase {
+
+  /**
+   * Some metrics will only produce warnings (never errors) on very small change sets.
+   */
+  private static final Collection<String> METRICS_THAT_CAN_ONLY_PRODUCE_WARNINGS_ON_SMALL_CHANGESETS = asList(
+    CoreMetrics.NEW_COVERAGE_KEY,
+    CoreMetrics.NEW_LINE_COVERAGE_KEY,
+    CoreMetrics.NEW_BRANCH_COVERAGE_KEY,
+    CoreMetrics.NEW_DUPLICATED_LINES_DENSITY_KEY,
+    CoreMetrics.NEW_DUPLICATED_LINES_KEY,
+    CoreMetrics.NEW_BLOCKS_DUPLICATED_KEY
+  );
+  private static final int MAXIMUM_NEW_LINES_FOR_SMALL_CHANGESETS = 19;
+
+  private final MeasureRepository measureRepository;
+  private final MetricRepository metricRepository;
+
+  public SmallChangesetQualityGateSpecialCase(MeasureRepository measureRepository, MetricRepository metricRepository) {
+    this.measureRepository = measureRepository;
+    this.metricRepository = metricRepository;
+  }
+
+  QualityGateMeasuresStep.MetricEvaluationResult applyIfNeeded(Component project, @Nullable QualityGateMeasuresStep.MetricEvaluationResult metricEvaluationResult) {
+    if (metricEvaluationResult == null) {
+      return metricEvaluationResult;
+    }
+    if (metricEvaluationResult.evaluationResult.getLevel() == Measure.Level.OK) {
+      return metricEvaluationResult;
+    }
+    if (!METRICS_THAT_CAN_ONLY_PRODUCE_WARNINGS_ON_SMALL_CHANGESETS.contains(metricEvaluationResult.condition.getMetric().getKey())) {
+      return metricEvaluationResult;
+    }
+    if (!isSmallChangeset(project)) {
+      return metricEvaluationResult;
+    }
+    return calculateModifiedResult(metricEvaluationResult);
+  }
+
+  private QualityGateMeasuresStep.MetricEvaluationResult calculateModifiedResult(@Nullable QualityGateMeasuresStep.MetricEvaluationResult metricEvaluationResult) {
+    return new QualityGateMeasuresStep.MetricEvaluationResult(
+      new EvaluationResult(Measure.Level.OK, metricEvaluationResult.evaluationResult.getValue()), metricEvaluationResult.condition);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') A "NullPointerException" could be thrown; "metricEvaluationResult" is nullable here. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259)